plugin.py

```diff
@@ -185,9 +186,29 @@ def find_package_dependency_dirs(self, py_ver: Tuple[int, int] = (3, 3)) -> List
         return [path for path in dep_dirs if os.path.isdir(path)]
 
     @classmethod
-    def resolve_python_path_from_venv(
-        cls, settings: DottedDict, workspace_folders: List[WorkspaceFolder]
-    ) -> Optional[str]:
+    def python_path(cls, settings: DottedDict, workspace_folders: List[WorkspaceFolder]) -> Optional[str]:
+        python_path = settings.get("python.pythonPath")
+        if python_path:
+            return python_path
+
+        if workspace_folders:
+            workspace_folder = os.path.abspath(workspace_folders[0].path)
+
+            while True:
+                python_path = cls.python_path_from_venv(workspace_folder)
+                if python_path:
+                    return python_path
+
+                parent = os.path.dirname(workspace_folder)
+                if workspace_folder != parent:
+                    workspace_folder = parent
+                else:
+                    break
+
+        return shutil.which("python")
```

Comment:
It's not entirely clear to me if there is gonna be any practical different in how pyright handles full binary path compared to just `python` that we've returned before.

It has some special logic for handling `python` and `python3` vs. full paths...

https://github.com/microsoft/pyright/blob/9eeeaf46c3161b42b080fc930027af02c6ad9c3b/packages/pyright-internal/src/server.ts#L98-L103

Reply:
I added it to improve the log output (Line 74-75).

Reply:
That makes sense but I'm a little bit concerned that it might result in different behavior in pyright.

Though I suspect that I'm worrying too much in that case. We can try it and revert if necessary.

Reply:
I actually have one more concern. I remember something about `shutil.which` not working correctly on Windows when checking symlinks. `python` is typically a symlink on Unix but I guess not on Windows so it should be fine.

Reply:
> I remember something about `shutil.which` not working correctly on Windows when checking symlinks. `python` is typically a symlink on Unix but I guess not on Windows so it should be fine.

Not sure whether you mean this:

```
>>> import shutil
>>> shutil.which ('wt')
'C:\\Users\\jfcherng\\AppData\\Local\\Microsoft\\WindowsApps\\wt.EXE'
Python 3.3>>> import shutil
Python 3.3>>> shutil.which ('wt')
```

It works in py38 but not in py33 for some special cases.

P.s. `wt` is an alias of Windows Terminal. If the user installs Python via Microsoft Store, it may hit this? I am not sure.
![image](https://user-images.githubusercontent.com/6594915/224935787-47d14a8d-2b0c-4b79-ae17-4ba9d0b48e04.png)

Reply:
Yep, that's the bug I had in mind.

If there can be cases where python.exe/python3.exe is a alias/link/junction (or whatever those are called on Windows) then it can be a problem indeed.

Pyright resolving likely doesn't have this issue so deferring to it would likely be better.

Reply:
Should I add a fallback to `python` if `shutil.which("python")` fails or drop `shutil` completely?

Reply:
adding `or "python"` sounds good for me if your initial intension is to get full path.

Reply:
Done.